flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/data/RowDataNestProjection.java

```diff
@@ -0,0 +1,281 @@
+            projectedMap.valueType().equals(originalMap.valueType());
+        Preconditions.checkArgument(keyProjectable && valueProjectable,
+            "Cannot project a partial map key or value with non-primitive type. Trying to project <%s> out of <%s>",
+            projectField, rowField);
+
+        return RowData.createFieldGetter(rowType.getTypeAt(rowTypePosition),
+            projectFieldsType.fields().indexOf(projectField));
+
+      case LIST:
+        Types.ListType projectedList = projectField.type().asListType();
+        Types.ListType originalList = rowField.type().asListType();
+
+        boolean elementProjectable = !projectedList.elementType().isNestedType() ||
+            projectedList.elementType().equals(originalList.elementType());
+        Preconditions.checkArgument(elementProjectable,
+            "Cannot project a partial list element with non-primitive type. Trying to project <%s> out of <%s>",
+            projectField, rowField);
+
+        return RowData.createFieldGetter(rowType.getTypeAt(rowTypePosition),
+            projectFieldsType.fields().indexOf(projectField));
+      default:
+        return RowData.createFieldGetter(rowType.getTypeAt(rowTypePosition),
+            projectFieldsType.fields().indexOf(projectField));
+    }
+  }
+
+  public static RowDataNestProjection create(Schema schema, Schema projectedSchema, int[][] projectedFields) {
+    return RowDataNestProjection.create(FlinkSchemaUtil.convert(schema), schema.asStruct(), schema.asStruct(),
+        projectedSchema.asStruct(),
+        projectedFields, true);
+  }
+
+  public static RowDataNestProjection create(RowType rowType,
+                                             Types.StructType schema,
+                                             Types.StructType rowStructType,
+                                             Types.StructType projectedSchema,
+                                             int[][] projectedFields,
+                                             boolean nestFlat) {
+    return new RowDataNestProjection(rowType, schema, rowStructType, projectedSchema, projectedFields, nestFlat);
+  }
+
+  public static RowDataNestProjection create(RowType rowType, Types.StructType schema, Types.StructType projectedSchema,
+                                             int[][] projectedFields, boolean nestFlat) {
+    return new RowDataNestProjection(rowType, schema, projectedSchema, projectedFields, nestFlat);
+  }
+
+  public RowData wrap(RowData row) {
+    this.rowData = row;
+    return this;
+  }
+
+  private Object getValue(int pos) {
+    Object fieldValue = getters[pos].getFieldOrNull(rowData);
+    while (nestFlat && fieldValue != null && fieldValue.getClass().equals(RowDataNestProjection.class)) {
+      RowDataNestProjection rowDataNest = (RowDataNestProjection) fieldValue;
+      fieldValue = rowDataNest.getters[rowDataNest.projectedFields[0][0]].getFieldOrNull(rowDataNest);
+    }
+    return fieldValue;
```

Comment:
Here, the nested fields are drilled down to extract the fields that need to be obtained.